src/transport/http2.rs

```diff
@@ -115,83 +117,47 @@ impl<In: RpcMessage, Out: RpcMessage> Clone for ClientChannel<In, Out> {
     }
 }
 
-/// A channel using a hyper connection
-pub struct ServerChannel<In: RpcMessage, Out: RpcMessage>(
-    flume::Receiver<(flume::Receiver<In>, flume::Sender<Out>)>,
-);
+/// A server-side channel using a hyper connection
+///
+/// Each request made by the any client connection this channel will yield a `(recv, send)`
+/// pair which allows receiving the request and sending the response.  Both these are
+/// channels themselves to support streaming requests and responses.
+///
+/// Creating this spawns a tokio task which runs the server, once dropped this task is shut
+/// down: no new connections will be accepted and existing channels will stop.
+#[derive(Debug)]
+pub struct ServerChannel<In: RpcMessage, Out: RpcMessage> {
+    /// The channel.
+    channel: Receiver<(Receiver<In>, Sender<Out>)>,
+    /// The sender to stop the server.
+    ///
+    /// We never send anything over this really, simply dropping it makes the receiver
+    /// complete and will shut down the hyper server.
+    stop_tx: mpsc::Sender<()>,
+}
 
 impl<In: RpcMessage, Out: RpcMessage> ServerChannel<In, Out> {
-    /// create a server given a socket addr
-    pub fn new(
-        addr: &SocketAddr,
-    ) -> hyper::Result<(Self, impl Future<Output = result::Result<(), hyper::Error>>)> {
+    /// Creates a server listening on the [`SocketAddr`].
+    pub fn new(addr: &SocketAddr) -> hyper::Result<Self> {
```

Comment:
I'm almost tempted to rename this to `spawn` or `serve` or `listen` or something else than new.  I'm not really sure what the etiquette is about spawning stuff in new functions.

Reply:
It used to not spawn, but return a future that you were supposed to handle yourself. So I guess a renaming is appropriate.